File: sage/src/main/java/com/hartwig/hmftools/sage/coverage/ExonCoverage.java
package com.hartwig.hmftools.sage.coverage;

import static java.lang.String.format;

import static com.hartwig.hmftools.common.utils.sv.BaseRegion.positionsOverlap;

import java.util.List;

import com.hartwig.hmftools.common.genome.region.GenomeRegion;
import com.hartwig.hmftools.common.utils.VectorUtils;

import org.apache.commons.compress.utils.Lists;

public class ExonCoverage
{
    private final GenomeRegion mRegion;
    private final int mExonRank;
    private final int[] mBaseCoverage;

    public ExonCoverage(final GenomeRegion region, final int exonRank)
    {
        mRegion = region;
        mExonRank = exonRank;
        mBaseCoverage = new int[region.bases()];
    }

    public final GenomeRegion region() { return mRegion; }
    public int exonRank() { return mExonRank; }

    public int[] coverage() { return mBaseCoverage;}

    public String chromosome() { return mRegion.chromosome(); }
    public int start() { return mRegion.start(); }
    public int end() { return mRegion.end(); }

    public void processRead(int readStartPos, int readEndPos)
    {
        if(!positionsOverlap(readStartPos, readEndPos, mRegion.start(), mRegion.end()))
            return;

        int startPosition = Math.max(start(), readStartPos);
        int endPosition = Math.min(end(), readEndPos);

        int startIndex = index(startPosition);
        int endIndex = index(endPosition);

        synchronized (mBaseCoverage)
        {
            for(int i = startIndex; i <= endIndex; i++)
            {
                mBaseCoverage[i] += 1;
            }
        }
    }

    public double medianDepth()
    {
        List<Double> depths = Lists.newArrayList();

        for(int i = 0; i < mBaseCoverage.length; ++i)
        {
            VectorUtils.optimisedAdd(depths, mBaseCoverage[i], true);
        }

        int medianIndex = mBaseCoverage.length / 2;
        return depths.get(medianIndex);
    }

    private int index(int position) { return position - start(); }

    public String toString() { return format("rank(%d) region(%s)", mExonRank, mRegion.toString()); }
}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilac/coverage/ExonCoverage.kt
package com.hartwig.hmftools.lilac.coverage

class ExonCoverage(val exonSequence: String) {
    val coverage = Array(exonSequence.length) { 0 }

    fun length(): Int {
        return exonSequence.length
    }

    fun addCoverageFromStart(n: Int) {
        synchronized(coverage) {
            for (i in 0 until n) {
                coverage[i] += 1
            }
        }
    }

    fun addCoverageFromEnd(n: Int) {
        synchronized(coverage) {
            for (i in coverage.size - n until coverage.size) {
                coverage[i] += 1
            }
        }
    }

    fun addCompleteCoverage() {
        synchronized(coverage) {
            for (i in coverage.indices) {
                coverage[i] += 1
            }
        }
    }

    fun addCoverage(index: Int, length: Int) {
        synchronized(coverage) {
            for (i in 0 until length) {
                coverage[index + i] += 1
            }
        }
    }


    fun isCovered(): Boolean {
        return minCoverage() > 0
    }

    fun minCoverage(): Int {
        return coverage.min()!!
    }

    fun maxCoverage(): Int {
        return coverage.max()!!
    }
}